### DISCUSSION TITLE
Signals are troublesome with Typescript

### DISCUSSION THREAD

**Discussion by stefnotch** [solidjs/solid#1575] [Q&A] [answered]
Lots of typical Typescript code uses [union types](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types). For example, when a value can either be a number, or can be null. `number | null`.
Or it can be a discriminated union, see https://github.com/solidjs/solid/discussions/1575#discussioncomment-5086272

### Examples
Here is a very simplistic example
```typescript
let count = null as number | null;

const increment = () => { count = (count ?? 0) + 1; }; 

if(count != null) {
  // notice how Typescript is happy, because it knows that count is not null
  console.log("next value" + (count + 1));
}
```

Attempting to do the same using a signal and the `Show` tag is painful though. 

```typescript
export default function Counter() {
  const [count, setCount] = createSignal<number|null>(null);

  return (
    <button class="increment" onClick={() => setCount((count() ?? 0) + 1)}>
      Clicks: {count()}
      <Show when={count() != null}>
{/* Typescript complains here, because it cannot figure out that count() won't be null */}
        next value will be {count() + 1}
      </Show>
    </button>
  );
}
```

This is mostly a toy example to illustrate my point. (The documentation also uses that toy example https://www.solidjs.com/guides/typescript#control-flow-narrowing ) This one can be worked around, but I think it becomes very painful once one writes anything bigger. 
e.g. Someone else attempted to use discriminated unions and didn't find any good solutions https://github.com/solidjs/solid/discussions/1527#discussioncomment-4950856
e.g. Nested properties don't have an easy way of being dealt with.

### Why?

I feel like this is a Typescript limitation, made worse by SolidJS's signal API design choice. 

As in, when using props instead of signals, it might become possible to use type assertions to properly type the `Show` tag ( https://github.com/microsoft/TypeScript/issues/40562 ).

But, this is much harder with a signal. A signal has a getter, which simply a function. However, a function does not have to return the same value every time, which means that Typescript cannot narrow its return type.

e.g.
```typescript
  const [count, setCount] = createSignal<number|null>(null);

  if(count() != null) {
    console.log("next value" + (count() + 1));
  }
```

The workaround there is to introduce a local variable, but of course that doesn't work in JSX.

**Comment by stefnotch** [comment]
Regarding the Switch and Match, here's a simple example that doesn't appear to have a good workaround.

```typescript
type User =
  | { tag: "anonymous" }
  | { tag: "user"; name: string }
  | { tag: "admin"; email: string };

const ViewA = (props: User) => (
  <Switch>
    <Match when={props.tag === "anonymous"}>Hello stranger!</Match>
    <Match when={props.tag === "user"}>
{/* Typescript errors here */}
      Hello {props.name}
    </Match>
    <Match when={props.tag === "admin"}>
{/* and another Typescript error here */}
      Hello admin at {props.email}
    </Match>
  </Switch>
);
```

**Reply by stefnotch** [reply]
@ryansolid I do apologise for the ping, but I feel like it would be important to raise this issue as early as possible. After all, fixing potentially major issues early seems like the easiest option. Limitations like these are blocking me from using Solid with Typescript. 

Is there a plan on how to eventually deal with this? e.g.
- Opening an issue with the lovely Typescript team as early as possible
- Investigating different API design choices
- Having a Solid language server

**Reply by ryansolid** [reply, marked as answer]
1. I tried to bring up my concern around idepotent functions with TS people I knew and they basically told me it was highly unlikely they'd ever be able to implement that in a reasonable way. I brought up how `getter`s are basically broken the opposite way and they basically consider it a choose your battles thing. To be fair I'd still write an issue but I lack the TS knowledge to even write up what I want it to do. But if this can be fixed the rest can be improved.

2. Having TS change API is hard... Mostly that there are reasons for the API the way it is. Reasons why for composition I don't just disfavor `. value` because of some opinion. We've talked about this at length before. [I have a whole stream about it](https://www.youtube.com/watch?v=2fW6JaJHoCE). But short of it is we want to support functions as they are instead of forcing wrappers on people.

Similarly with JSX we want our control flow to be composable. I'm not sure any automatic analysis could ever handle:

```js
<MyIf condition={something()}>
  <MaybeShown />
</MyIf>
```
Now we support ternaries and I could see a reality where switch or native control flow inside JSX blocks could be processed but it would have limited application because it wouldn't be understood outside of the immediate scope. Things like [do expressions](https://github.com/tc39/proposal-do-expressions)  might make that possible but it would be limited.

The function form of `<Show>` and `<For>` are at least helpful. For is fine atleast. `<Show>` is still tricky. Switch/Match I suppose is a but harder to get the same benefit. I am open to designing some different control flow around Switch/Match. I think though for any of these to work the idepontent functions need to come first. The fact that Angular uses functions for Signals (again for good reason) too I'm hoping helps this.

3. I mean that's a whole other thing and I'm not sure short of specializing tags this even helps outside of like making our own JSX fork. It's also [an insane amount of work](https://www.youtube.com/watch?v=W9wfkP21c80).
-----------------

If someone wants to help frame a proposal to TS for what I'm calling idepotent functions I will push it through all my channels to draw attention to it.

**Comment by atk** [comment]
You can help yourself with a guard function:

```ts
const isDef = <T extends any>(accessor: Accessor<T | undefined>): accessor is Accessor<T> =>
  accessor() !== undefined;

if (isDef(count)) {
  // herein, count will be handled as () => number
}
```

**Reply by stefnotch** [reply]
That's a lovely approach when using Typescript, but it also doesn't work with JSX

https://playground.solidjs.com/anonymous/27a17e1f-f30f-45ac-8419-650e4849fb41

**Reply by stefnotch** [reply]
The very naive way of moving the check into Typescript code also doesn't quite seem to cut it
https://playground.solidjs.com/anonymous/09d3cf29-72e4-45af-9ada-0d0eee36a619

I really wonder how that could be improved/fixed.

(For the record, Vue.js went down the road of writing their own language server for templates. While that definitely lets them solve such issues, it's also a lot of work. And needs an IDE extension.)

**Reply by atk** [reply]
Obviously, Solid's JSX conversion is putting the different reactive parts of the JSX into different function scopes, so you need to check isDef(accessor) in the same {}-part that you are attempting to use the accessor.

**Reply by stefnotch** [reply]
That of course works, but then I'm missing out on both shortness *and Typescript's ability to correctly check my code*.

As in, when I write this
```typescript
return (
    <button class="increment" onClick={() => setCount((count() ?? 0) + 1)}>
      Clicks: {count()}
      <Show when={isDef(count)}>
        next value will be {isDef(count) ? (count() + 1) : never()}
      </Show>
    </button>
  );
```

Then Typescript cannot warn me if I mess up and remove the `<Show>`. This code is clearly unintentional, and in an ideal world, Typescript would be able to catch such "missing check" mistakes.
```typescript
return (
    <button class="increment" onClick={() => setCount((count() ?? 0) + 1)}>
      Clicks: {count()}
        next value will be {isDef(count) ? (count() + 1) : never()}
    </button>
  );
```
Full example https://playground.solidjs.com/anonymous/8f169498-a8b4-432f-9102-14f61ebf7be4

I should probably also mention that I'm perfectly fine if the answer to this is "Solidjs doesn't have a truly good solution for this yet".
I just really want to raise this issue as early as possible, so that the lovely team behind Solidjs understands this shortcoming and will eventually find a nice solution. :)

**Reply by atk** [reply]
The actual answer is "TypeScript doesn't have a truly good solution for this yet". There is no way to create a function type that will make TypeScript recognize that its return type cannot change inside a closure. The lovely team behind Solid.js is painfully aware of this shortcoming and hopes that the wonderful team behind TypeScript will come up with a solution eventually.

**Reply by flevi29** [reply]
Is there an issue about this or a discussion open on TypeScript? I couldn't find any. Are they aware of this issue? I'm not sure they are.
This issue has come up in Angular as well: https://github.com/angular/angular/issues/49161

**Reply by stefnotch** [reply]
I'm not aware of a relevant upstream Typescript issue. @flevi29 It would be lovely if you could open one, to make them aware of one current issue in the ecosystem.

Although, chances are that the only short term solution is to adopt Vue.js-like signals. That is, not being function based and instead having a `.value` getter and setter.

Long term solutions could have many shapes and forms.

**Reply by flevi29** [reply]
I'm not sure if TypeScript will ever come up with a solution to this. Getters and setters path might as well be non-null assertion path in terms of type safety, but I guess at least you have discriminated unions and more comfort. I don't get it, TypeScript is sometimes hell bent on type safety, and other times we get stuff like this:

```typescript
class Asd {
  number = 0;

  get value(): number | null {
    this.number += 1;
    const variant = this.number % 2;
    const retval = variant === 0 ? null : this.number;
    console.log("called", variant, retval)
    return retval;
  }
}

const asd = new Asd();

for (let i = 0; i < 10; i += 1) {
  if (asd.value !== null) {
    const wrong: number = asd.value;
    console.log(wrong);
  }
}
```

**Reply by stefnotch** [reply]
Turns out that I was wrong. Apparently even a world where Solid has an API with a `.value` getter and setter wouldn't work, due to how JSX works.

https://playground.solidjs.com/anonymous/e428e57a-1c0b-44fe-9450-f70eeacc57a7

**Reply by stefnotch** [reply]
It might just work in Angular-land though https://github.com/angular/angular/issues/49161#issuecomment-1988728595

**Reply by eneajaho** [reply]
> It might just work in Angular-land though [angular/angular#49161 (comment)](https://github.com/angular/angular/issues/49161#issuecomment-1988728595)

This should be solved in Typescript level (and not template level either jsx/ng templates), because we don't use signals in templates only, but also in typescript code, so having 2 different patterns would break the mental model that you can use signals the same  everywhere.

**Comment by devzzzero** [comment]
Hi, this is not an answer, but I guess another way in which typescript + preact + signals cause a bit of issue:
I'm tying to use a set of signals as a global state for my app, and then I try to take a subset of the signal's value's (via peek()) to  
decorate components that  are "past history" (i.e. the component render a prior version), but that causes headaches as well.

In the code below, the object "arg" is a "dict" (sorry, a python-ism for just a regular object with keys and values (not a JS Map!) that contains all of the global signals for my app.  I then try to have a function that gives a snapshot of the app state for rendering "past history" and this is where I ran into apparently unsolvable issue with typescript.

```
function toProps(arg) {
  if (isDict(arg) && ) {
    const rtn = {};
    for (const [k,v] of Object.entries(arg)) {
      rtn[k] = v.peek(); // <---    typescript error: unknown how to fix
    }
    return rtn;
  } else if (isArray(arg)) {
    const rtn = arg.map((v) => v.peek());
    return rtn;
  }
  throw new TypeError(`${arg} is not a container of signals`);
}
```
I tried all kinds of ways to get rid of that error. After about 5 hours, I gave up.

**Reply by stefnotch** [reply]
This seems unrelated to Solid.js. There are quite a few ways of solving this one in particular, and without knowing more about the code, it's not easy to suggest one.

For example, you could do `rtn[k] = (v as any).peek();` to silence that particular error. This does slightly go against the whole point of Typescript though.

**Comment by kettanaito** [comment]
I've found a bit of success achieving proper non-nullable resources in my Solid components when using routing (TanStack Router in this case). The idea is to validate the path parameter that points to a resource before rendering the component and then rely on it being non-nullable after the check.

```ts
export const Route = createFileRoute('/resources/$resourceId')({
  component: ResourceComponent,
  loader({ params }) {
    const resource = findResource(params.resourceId)

    if (!resource) {
      throw redirect({ to: '/' })
    }

    return { resource }
  },
})

function ResourceComponent() {
  const data = Route.useLoaderData()
  const resource = () => data().resource
  //    ^ `resource` is a non-nullable T

  return <p>{resource().title}</p>
}
```

**Comment by kbrilla** [comment]
Hi @ryansolid , I'm working on a fix for that, not ready to bring it up to the TS people, but I would love some comments from the SolidJS team if possible https://github.com/kbrilla/typescript-go/pull/2   I can't say enough how WIP that is, but well, it does work :)  Of course, I admit that it is written by Opus 4.6 in collaboration with GPT 5.3 Codex, not me personally, but well under my watch. Bdw. doing that in ts-go as TS stopped approving PR to TS repo, but I think I could menage to rewrite it to non Go TS as well 

<img width="919" height="520" alt="image" src="https://github.com/user-attachments/assets/e411c2f4-1a8d-4431-bdc7-be81de66f9cb" />

**Reply by kbrilla** [reply]
@stefnotch https://github.com/kbrilla/solidjs-stable-mutator-demo :) and https://stackblitz.com/github/kbrilla/solidjs-stable-mutator-demo  - still updating vsix to work better, but well, here we go :)  run run.sh before to get correct binaires 
<img width="807" height="98" alt="image" src="https://github.com/user-attachments/assets/79a0a966-3e9f-4f6f-bf90-87cdfc184dc4" />
